src/main/lrsql/ops/query/reaction.clj

```diff
@@ -0,0 +1,163 @@
+(ns lrsql.ops.query.reaction
+  (:require [clojure.spec.alpha :as s]
+            [lrsql.backend.protocol :as bp]
+            [lrsql.spec.common :refer [transaction?]]
+            [lrsql.spec.reaction :as rs]
+            [lrsql.util.reaction :as ru]
+            [lrsql.ops.util.reaction :as ur]
+            [lrsql.util :as u]
+            [xapi-schema.spec :as xs]
+            [clojure.tools.logging :as log]))
+
+(s/fdef query-all-reactions
+  :args (s/cat :bx rs/reaction-backend?
+               :tx transaction?)
+  :ret rs/query-all-reactions-ret-spec)
+
+(defn query-all-reactions
+  "Return all reactions."
+  [bk tx]
+  (mapv
+   (fn [{:keys [id
+                title
+                ruleset
+                active
+                created
+                modified
+                error]}]
+     {:id       id
+      :title    title
+      :ruleset  (ru/stringify-template ruleset)
+      :active   active
+      :created  created
+      :modified modified
+      :error    error})
+   (bp/-query-all-reactions bk tx)))
+
+(s/fdef query-statement-reactions
+  :args (s/cat :bk rs/reaction-backend?
+               :tx transaction?
+               :input rs/query-statement-reactions-input-spec)
+  :ret rs/query-statement-reactions-ret-spec)
+
+(defn- reaction-error-response
+  "Return a properly formatted reaction error response"
+  [reaction-id trigger-id error-type error-message]
+  {:reaction-id reaction-id
+   :trigger-id  trigger-id
+   :error       {:type    error-type
+                 :message error-message}})
+
+(defn- reaction-query
+  [bk tx ruleset reaction-id trigger-id trigger-stored statement-identity]
+  (try
+    [true (ur/query-reaction
+           bk
+           tx
+           {:ruleset            ruleset
+            :trigger-id         trigger-id
+            :trigger-stored     trigger-stored
+            :statement-identity statement-identity})]
+    (catch Exception ex
+      (log/errorf
+       ex
+       "Reaction Query Error - Reaction ID: %s"
+       reaction-id)
+      [false (reaction-error-response
+              reaction-id
+              trigger-id
+              "ReactionQueryError"
+              (ex-message ex))])))
+
+(defn- generate-statement
+  [reaction-id trigger-id ruleset-match template]
+  (try
+    [true (ru/generate-statement
+           ruleset-match
+           template)]
+    (catch Exception ex
+      (log/errorf
+       ex
+       "Reaction Template Error - Reaction ID: %s"
+       reaction-id)
+      [false (reaction-error-response
+              reaction-id
+              trigger-id
+              "ReactionTemplateError"
+              (ex-message ex))])))
+
+(defn query-statement-reactions
```

Comment:
I would like it if this function was broken up into multiple pieces. It's difficult to navigate with the sheer length and multiple levels of nesting.

Reply:
https://github.com/yetanalytics/lrsql/pull/306/commits/c9dacb177aa2c65079aafd469c53e789c7a56f2e